### ISSUE TITLE
In the documentation create rules list with descriptions

### ISSUE THREAD

**Issue by MNF** [nsubstitute/NSubstitute.Analyzers#121]
Currently in https://github.com/nsubstitute/NSubstitute.Analyzers/tree/master/documentation rules list refer to github folder https://github.com/nsubstitute/NSubstitute.Analyzers/tree/master/documentation/rules that has rule code (e.g. NS2001.md) without any meaningful description and reader have to open each file to understand what the rule is about.
It will be better to create In the documentation a file with the list of rule  descriptions to have on overview of the rules.
E.g.

```
| ID       | Category      | Description |
| ———————— | ——————————-—— | ——————-—————-|
| NS2001   | Substitute    |  A violation of this rule occurs when Substitute.ForPartsOf or |
|          | creation.     | Substitute.For is used for substituting type which does not |
|          |               | expose public  or protected constructor |
```

**Comment by dtchepak** [comment]
Thanks for raising this @MNF. I've been thinking of maybe trying to do this dynamically with some JS, but I'm worried that may be a bit unreliable. Maintaining a separate page that duplicates the information on the sub-pages risks it becoming out-dated (also unreliable). Alternatively we could generate it as part of the build and check it in if it changes (also unreliable?). Or another option is to generate this list as part of NSubstitute's website build (add to [this page](https://nsubstitute.github.io/help/nsubstitute-analysers/)?).

Any thoughts on these options? (Or suggestions for other options?)

**Comment by MNF** [comment]
As the structure of NS****.md files seems consistent, it should be not hard to create utility to loop for all files in the folder and extract ID ,Category, Description and save them in the .md/.html table( and report any errors if structure of the file is  inconsistent and corresponding field not found)
Architecturally I prefer to generate the file during website build rather than each repository build, but I am not familiar, how easy is to invoke extra step during website build.

**Comment by tpodolak** [comment]
I don't think we need to write a generator for the rules overview. In my opinion rule documentation will not be changed too often, so we can keep the data in sync manually as long as we have a test which will verify the content of both documents against each other. I would also change the @MNF proposed table to sth like this
### Rules

| ID       | Category      | Cause |
|---|---|---|
| [NS1000](rules/NS1000)  | Non-substitutable member | Substituting for non-virtual member of a class. |
| [NS1001](rules/NS1001)  | Non-substitutable member | Checking received calls for non-virtual member of a class.  |
| [NS1002](rules/NS1002)  | Non-substitutable member | Substituting for non-virtual member of a class. |
| [NS1003](rules/NS1003)  | Non-substitutable member | Substituting for an internal member of a class without proxies having visibility into internal members.  |
| [NS2000](rules/NS2000)  | Substitute creation | Substitute.ForPartsOf used with interface or delegate. |
| [NS2001](rules/NS2001)  | Substitute creation | NSubstitute used with class which does not expose public or protected constructor. |
| [NS2002](rules/NS2002)  | Substitute creation | NSubstitute used with class which does not expose parameterless constructor. |
| [NS2003](rules/NS2003)  | Substitute creation | NSubstitute used with internal type. |
| [NS2004](rules/NS2004)  | Substitute creation | Substituting for type by passing wrong constructor arguments. |
| [NS2005](rules/NS2005)  | Substitute creation | Substituting for multiple classes. |
| [NS2006](rules/NS2006)  | Substitute creation | Substituting for interface and passing arguments. |
| [NS2007](rules/NS2007)  | Substitute creation | Substituting for delegate and passing arguments. |
| [NS3000](rules/NS3000)  | Argument specification | Accessing call arguments out of the bounds of method arguments. |
| [NS3001](rules/NS3001)  | Argument specification | Casting call argument at given position to different type than type specified in a method. |
| [NS3002](rules/NS3002)  | Argument specification | Accessing call argument by type which is not present in invocation. |
| [NS3003](rules/NS3003)  | Argument specification | Accessing call argument by type which is used multiple times in invocation. |
| [NS3004](rules/NS3004)  | Argument specification | Assigning call argument with type which is not the same as method argument type. |
| [NS3005](rules/NS3005)  | Argument specification | Assigning call argument which is not ref nor out argument. |
| [NS3006](rules/NS3006)  | Argument specification | Conflicting assignments to out/ref arguments. |
| [NS5000](rules/NS5000)  | Usage  | Checking received calls without specifying member. |

So instead of **Description**, I would show **Cause** - as a description tends to be long and detailed  explanation of the rule. Thoughts @MNF @dtchepak

**Comment by MNF** [comment]
Table looks good for me.

**Comment by dtchepak** [comment]
I've opened a PR for this: https://github.com/nsubstitute/NSubstitute.Analyzers/pull/122

[Sample output](https://github.com/dtchepak/NSubstitute.Analyzers/blob/GH121-tableofcontents/documentation/rules/README.md)

**Comment by tpodolak** [comment]
@dtchepak @MNF  closing as the documentation was updated with rules summary 
Closed by #122 and #123
